### PR TITLE
Exam bank interface

### PR DESCRIPTION
**Pull request by ad-world**

### CODE REVIEW

File: public/assets/ts/admin/exam-bank.ts [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+          "$SOLUTION_HIDDEN$",
+          exam.solutionFile?.includes("-hidden") ? "True" : "False"
+        );
+
+      newRow.setAttribute("data-course-dept", exam.department);
+      newRow.setAttribute("data-course-code", exam.courseCode);
+
+      const hideExamButton = newRow.getElementsByClassName("exam-actions")[0]
+        .children[1] as HTMLButtonElement;
+      const hideExamSolutionButton = newRow.getElementsByClassName(
```
Comment (some-molasses) [inline public/assets/ts/admin/exam-bank.ts:134]: Can we prevent the solution buttons from displaying when there's no solution file?

Similarly, can we prevent the exam buttons from displaying when there's a solution file without an exam file?

File: server/api/controllers/exam-bank-controller.ts [hunk 2]
```diff
@@ -1,80 +1,170 @@
...
+    res.send({ status: "success", errors: [] });
+  }
+
+  async hideExamFile(examName: string): Promise<void> {
     const currentUrl = `public/exams/${examName}.pdf`;
     const newUrl = `public/exams/${examName}-hidden.pdf`;
 
-    if (!fs.existsSync(currentUrl)) {
+    if (!existsSync(currentUrl)) {
       throw new Error("Exam not found");
```
Comment (some-molasses) [inline server/api/controllers/exam-bank-controller.ts:146]: This is throwing an error that crashes the site when you try to hide the same exam twice in a row - want me to take a look? I think I may have caused it originally lol

**Review by some-molasses — COMMENTED** [review]
thanks for taking this on!

File: server/api/admin/index.ts [hunk 3]
```diff
@@ -60,22 +53,24 @@ router.post("/data", validate, (req: Request, res: Response) => {
 });
 
 router.post("/exams/rebuild", (_req: Request, res: Response) => {
-  ExamBankController.rewriteFile();
+  new ExamBankController().rewriteFileJson();
   res.status(201).send();
 });
 
+router.post("/exams/upload", (req: Request, res: Response) => {
+  new ExamBankController().uploadFiles(req, res);
```
[the remark this answers is not in the document]
Comment (ad-world) [reply server/api/admin/index.ts:61]: I can kinda see your point but why wouldn't this work if we created a new queue each time?

**Comment by some-molasses** [comment]
closed because stale 😔

**Pull request closed without merging**